### PR TITLE
[InMemoryDataset redesign] EntireChunksMapper

### PR DESCRIPTION
**Pull request by crusaderky**
- Part of #386
Blocked by and incorporates:
- #388 
- #373 
- #390

Add new class EntireChunksMapper, which reamains dormant until a later PR.

### CODE REVIEW

File: versioned_hdf5/tests/test_subchunk_map.py [hunk 1]
```diff
@@ -409,6 +410,55 @@ def test_read_many_slices_param_nd(args):
...
+    sel_chunks = all_chunks[orig_mapper.chunks_indexer()]
+    np.testing.assert_array_equal(
+        all_chunks[entire_mapper.chunks_indexer()], sel_chunks
+    )
+    np.testing.assert_array_equal(
+        sel_chunks[entire_mapper.whole_chunks_idxidx()], sel_chunks
+    )
+
+    # Test read_many_slices_params()
+    entire_mapper = EntireChunksMapper(orig_mapper)
```
Comment (peytondmurray) [inline versioned_hdf5/tests/test_subchunk_map.py:437]: I think you can drop this, the state of `entire_mapper` has not changed, right?

**Review by peytondmurray — APPROVED** [review]
Looks good - two minor comments, otherwise good to go :ship:

File: versioned_hdf5/subchunk_map.py [hunk 2]
```diff
@@ -514,6 +514,52 @@ def value_view_idx(self) -> slice | None:
...
 
+@cython.cclass
+class EntireChunksMapper(BasicChunkMapper):
+    """Special mapper that selects all points on the chunks selected by another mapper.
+
+    This is used to load the entire chunk for the purpose of caching when the
+    actual selection may only target part of it.
+    """
+
+    _chunks_indexer: slice | NDArray[np.intp]
```
[the remark this answers is not in the document]
Comment (crusaderky) [reply versioned_hdf5/subchunk_map.py:525]: This is a normal instance attribute. It's the _output_ of the `chunk_indexer` method of the original indexer.

File: versioned_hdf5/tests/test_subchunk_map.py (hunk 1 above)
Comment (crusaderky) [reply versioned_hdf5/tests/test_subchunk_map.py]: Good catch!

**Pull request merged**